### PR TITLE
updating boost.histogram

### PR DESCRIPTION
**Pull request by HDembinski**
* updates boost.histogram to develop
* updates the reduce interface to match boost.histogram
* added a lot more tests (test coverage for reduce and friends had large holes)

### CODE REVIEW

**Comment by henryiii** [comment]
This looks great!

File: src/register_algorithm.cpp [hunk 1]
```diff
@@ -9,46 +9,48 @@
...
+                                   "end={3}{4}, mode={5}))")
+                        .format(suffix,
+                                start,
+                                self.begin.index,
+                                self.end.index,
+                                merge,
+                                self.crop);
+                } else {
+                    return py::
+                        str("reduce_command(shrink{0}({1}, lower={2}, upper={3}{4}))")
```
Comment (henryiii) [inline src/register_algorithm.cpp]: Much cleaner

**Comment by henryiii** [comment]
Let's go ahead and merge so I can fix #332 and #331 on top (should be quick and easy). Ping me or use WIP/Draft mode if something is not ready!

**Comment by HDembinski** [comment]
You can also assgin #332 and #331 to me if you work on other things, there are easy fixes and this week I have some time to do some unpaid work on software.

**Comment by henryiii** [comment]
Okay, I'll leave it to you then, that sounds fine. I sent an email with a possible plan for the next release, let me know what you think.

**Comment by henryiii** [comment]
My [weekly CI develop update job](https://dev.azure.com/scikit-hep/boost-histogram/_build?definitionId=7&_a=summary) had already told me that work needed to be done in order to adapt to the latest Boost.Histogram develop changes. Nice to see it working. :)

I think we can simplify the indexing code a bit with these changes, will have to look into it in the near future.

**Pull request merged**